model.py

```diff
@@ -1,103 +1,1222 @@
+        return self.__get_descendant_parameter_vector__()
+
+    @par_val_list.setter
+    def par_val_list(self, par_val_list):
+        if not len(par_val_list) >= len(self.par_val_list):
+            raise ValueError('Fail on {}'.format(self.name))
+        self.__set_parameter_vector__(par_val_list)
+
+    @property
+    def par_dict(self):
+        '''A dict of the Param objects ABOVE! this node'''
+        return {key: val for key, val in
+                zip(self.__get_inherited_parameter_names__(),
+                    self.__get_inherited_parameter_vector__())}
+
+    @property
+    def is_root(self):
+        '''True if I have no parents'''
+        return self.parent is None
+
+    @property
+    def is_leaf(self):
+        '''True if I have no children'''
+        return len(self.children) == 0
+
+    # Diagnostic methods
+    @property
+    def structure(self):
+        '''Return the tree structure below me as a str, generated from nx.'''
+        self.create_tree()
+
+        return nx.readwrite.tree_data(self.nx_graph, self.name)
+
+    def report_relatives(self):
+        '''This is a pretty crappy, inflexible way of doing this. Can I
+        come up with a nicer, perhaps recursive way of it?'''
+        print("Reporting family tree of {}:".format(self.name))
+        try:
+            parent = self.parent.name
+        except:
+            parent = 'None'
+        print("    Parent: {}".format(parent))
+        print("    Children:")
+        for child in self.children:
+            print("      {}".format(child.name))
+            for grandchild in child.children:
+                print("       - {}".format(grandchild.name))
+
+    def report(self, also_relatives=True):
+        if also_relatives:
+            self.report_relatives()
+        print("  Parameter vector, and labels:")
+        for par, val in zip(self.par_names, self.par_val_list):
+            print("  {:>10} = {:<.3f}".format(par, val))
+        print("\n")
+
+    def create_tree(self, G=None, called=True):
+        '''Construct a tree node graph of the model structure.
+        Start from the called tier, and work down from there.'''
+        if called:
+            G = nx.DiGraph()
+            G.add_node(self.name)
+
+            for child in self.children:
+                # Add the child's children to the graph
+                G = child.create_tree(G, called=False)
+                # Connect myself to the child
+                G.add_edge(self.name, child.name)
+
+        else:
+            # Add myself to the graph
+            G.add_node(self.name)
+            for child in self.children:
+                # Add my child as a node, and connect it to me
+                G.add_node(child.name)
+                G.add_edge(self.name, child.name)
+                G = child.create_tree(G, called=False)
+
+            return G
+
+        self.nx_graph = G
+        return G
+
+    def plot_data(self, save=False):
```

Comment:
Here is an example where we should think a little about generality. Should a leaf node of a model always carry around its data? I'm not sure. This might be an ```Eclipse``` specific method that shouldn't be on the abstract Model class

Reply:
The reason it's on the abstract class is so that calling node.plot_data() will fall down the tree and plot all models descended from that node. I could copy-paste the function to the subclasses, but repeating code stresses me out... Perhaps a general plot_model(), that gets overwritten on the Eclipses to plot their specific data storage case? 

Alternatively, we could do something with the search_Node() function, like

```
for iecl in range(necl):
    eclipse = model.search_Node('Eclipse', str(iecl))
    eclipse.plot_data(save=True/False, publishable=True/False)
```

And just have the plot_data() function on the Eclipse subclass? I've used a similar thing elsewhere. Could even have the above be a function on the LCModel subclass so that calling model.plot_data() still works the same.

Reply:
I've done this now. Plotting is handled on the LCModel and the Eclipse levels. calling LCModel.plot_data() searches the tree for all eclipse nodes, and calls each of their data plotting functions.